### PR TITLE
i18n (internationalization)

### PR DESCRIPTION
**Pull request by lexjacobs**
This pull request covers the result of research and implementation of a JavaScript implementation of the gnu gettext i18n standard. A library called Jed.js was found and implemented on this branch and a very basic sample English/Japanese translation example was implemented.

Prior to merging this branch, all keys will be removed from the Japanese i18n domain, which will cause all translated strings to pass through as their original English versions. When the .po file is prepared by an actual translator, the translations can be reimplemented.

The implementation of this solution is limited, pending review and feedback. In order for a full site-wide translation, the `goldstone.translate(string)` scheme devised here will need to be implemented site-wide. It has been implemented enough to see the results, and offer feedback on the implementation.

In order to try this out, just pull this branch, head to `http://localhost:8000/#settings` and toggle the drop-down selector for 'language'. Some strings on the settings page, and some base.html template text will be translated to Japanese characters (obtained from google translate, and not necessarily correct).

You can also see the results of a translated heading on `http://localhost:8000/#metrics/api_perf` in the Nova API Perf header.

Of note: this is not a fully dynamic solution. .po files obtained from translators will have to be parsed with the new grunt tool added to the dev tools (and documented in Gruntfile.js), and manually added to the  `combinedPoJsonFiles` key in i18nModel.js .

Again, prior to merge, the translations obtained from google translate will need to be stripped out of the .po file, and Jed.js json object.

### CODE REVIEW

**Comment by jxstanford** [comment]
> Of note: this is not a fully dynamic solution. .po files obtained from translators will have to be parsed with the new grunt tool added to the dev tools (and documented in Gruntfile.js), and manually added to the combinedPoJsonFiles key in i18nModel.js .

Can you describe this process?  Is the manual part temporary?  Could we have the grunt task do the conversion to a json file, then have i18nModels.js read in the object from a file rather than embedding the object in code?

File: client/js/models/i18nModel.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    },
+
+    setTranslationObject: function() {
+        goldstone.translationObject = new Jed(this.combinedPoJsonFiles);
+        this.setTranslationFunction();
+    },
+
+    /*
+    these are the function signatures for the api returned by
+    createing a new Jed object:
```
Comment (stugots) [inline client/js/models/i18nModel.js]: creating

File: goldstone/static/bundle/bundle.js [hunk 2]
```diff
@@ -2737,6 +2735,158 @@ var GoldstoneColors = GoldstoneBaseModel.extend({
...
+    },
+
+    setTranslationObject: function() {
+        goldstone.translationObject = new Jed(this.combinedPoJsonFiles);
+        this.setTranslationFunction();
+    },
+
+    /*
+    these are the function signatures for the api returned by
+    createing a new Jed object:
```
Comment (stugots) [inline goldstone/static/bundle/bundle.js]: creating

**Comment by stugots** [comment]
Are any strings returned by any server API, including opentrail, displayed as-is by the client?

**Comment by lexjacobs** [comment]
@stugots it will depend on the i18n spec for goldstone server and addons as far as what is internationalized. When a string is run through the JavaScript implementation of gettext, if there is no translation key found, the original string is returned. So, any string this is either not wrapped in a translation function, and any string that is not matched with a translation, will be displayed as-is.

**Comment by lexjacobs** [comment]
@jxstanford I found an option in the `po2json` tool that will return a single file such as:

```
{
    "en": {
        "domain": "messages",
        "locale_data": {
            "messages": {
                "": {
                    "domain": "messages",
                    "plural_forms": "nplurals=2; plural=(n != 1);",
                    "lang": "en"
                },
                "Nova API Performance": [null, ""],
                "Neutron API Performance": [null, ""],
                . . .
            }
        }
    },
    "ja": {
        "domain": "messages",
        "locale_data": {
            "messages": {
                "": {
                    "domain": "messages",
                    "plural_forms": "nplurals=1; plural=0;",
                    "lang": "ja"
                },
                "Nova API Performance": [null, "新星のAPIのパフォーマンス"],
                "Neutron API Performance": [null, "中性子のAPIパフォーマンス"],
                . . .
            }
        }
    }
}
```

I can set the output location of the .po > json converter to a static folder in goldstone. If that json object were served by django on the base html template as
`var translations = {{i18nblob}}`
then I can modify i18nModels.js to construct a Jed.js-compatible object for instantiating the Jed translation object.

**Comment by lexjacobs** [comment]
@jxstanford @stugots 
Dynamic solution based on existing .po files has been implemented.
Please let me know if you have any further comments/questions.
Prior to merging branch, Japanese .po file will be removed from repo.

**Comment by jxstanford** [comment]
Please prepare the necessary instructions to give to a translator and include it in an I18N.md file under docs folder.  Doc should include links to relevant files and explicit instructions on how to add a new language and translations.

Ensure that all static text (menus, info pop ups, help page, etc.) is localized.

**Pull request merged**